src/lapack/backends/rocsolver/rocsolver_lapack.cpp

```diff
@@ -0,0 +1,3218 @@
+                         oneapi::mkl::transpose trans, std::int64_t m, std::int64_t n, T *a,
+                         std::int64_t lda, T *tau, T *c, std::int64_t ldc, T *scratchpad,
+                         std::int64_t scratchpad_size,
+                         const std::vector<sycl::event> &dependencies) {
+    using rocmDataType = typename RocmEquivalentType<T>::Type;
+    overflow_check(m, n, lda, ldc, scratchpad_size);
+    auto done = queue.submit([&](sycl::handler &cgh) {
+        int64_t num_events = dependencies.size();
+        for (int64_t i = 0; i < num_events; i++) {
+            cgh.depends_on(dependencies[i]);
+        }
+        onemkl_rocsolver_host_task(cgh, queue, [=](RocsolverScopedContextHandler &sc) {
+            auto handle = sc.get_handle(queue);
+            auto a_ = reinterpret_cast<rocmDataType *>(a);
+            auto tau_ = reinterpret_cast<rocmDataType *>(tau);
+            auto c_ = reinterpret_cast<rocmDataType *>(c);
+            auto scratch_ = reinterpret_cast<rocmDataType *>(scratchpad);
+            rocblas_status err;
+            ROCSOLVER_ERROR_FUNC_T(func_name, func, err, handle, get_rocblas_side_mode(side),
+                                   get_rocblas_fill_mode(uplo), get_rocblas_operation(trans), m, n,
+                                   a_, lda, tau_, c_, ldc);
+        });
+    });
+    return done;
+}
+
+#define UNMTR_LAUNCHER_USM(TYPE, ROCSOLVER_ROUTINE)                                             \
+    sycl::event unmtr(sycl::queue &queue, oneapi::mkl::side side, oneapi::mkl::uplo uplo,       \
+                      oneapi::mkl::transpose trans, std::int64_t m, std::int64_t n, TYPE *a,    \
+                      std::int64_t lda, TYPE *tau, TYPE *c, std::int64_t ldc, TYPE *scratchpad, \
+                      std::int64_t scratchpad_size,                                             \
+                      const std::vector<sycl::event> &dependencies) {                           \
+        return unmtr(#ROCSOLVER_ROUTINE, ROCSOLVER_ROUTINE, queue, side, uplo, trans, m, n, a,  \
+                     lda, tau, c, ldc, scratchpad, scratchpad_size, dependencies);              \
+    }
+
+UNMTR_LAUNCHER_USM(std::complex<float>, rocsolver_cunmtr)
+UNMTR_LAUNCHER_USM(std::complex<double>, rocsolver_zunmtr)
+
+#undef UNMTR_LAUNCHER_USM
+
+// SCRATCHPAD APIs
+
+template <typename Func>
+inline void gebrd_scratchpad_size(const char *func_name, Func func, sycl::queue &queue,
+                                  std::int64_t m, std::int64_t n, std::int64_t lda,
+                                  int *scratch_size) {
+    queue.submit([&](sycl::handler &cgh) {
+        onemkl_rocsolver_host_task(cgh, queue, [=](RocsolverScopedContextHandler &sc) {
+            auto handle = sc.get_handle(queue);
+            rocblas_status err;
+            ROCSOLVER_ERROR_FUNC_T(func_name, func, err, handle, m, n, scratch_size);
+        });
+    });
+}
+
+// TBD: rocSolver doesn't need/support scratchpad_size
```

Comment:
we are not sure what to return as scratchpad size because rocsolver doesn't use scratchpad., Please suggest what should be return here.? or how it should be handled.?

Reply:
I don't expect much trouble with returning 0. Users should be able to allocate as normal since SYCL 2020 spec now allows zero ranged buffers and usm allocations should just return a nullptr.

Reply:
Thanks for the suggestion @ericlars.

Reply:
> I don't expect much trouble with returning 0. Users should be able to allocate as normal since SYCL 2020 spec now allows zero ranged buffers and usm allocations should just return a nullptr.

@ericlars , When zero is returned from "get_scratchpad_size" APIs, the following error has been reported during testing for HIP backend.
"SYCL buffer size is zero. To create a device accessor, SYCL buffer size must be greater than zero."

Reply:
We've found that this issue was coming from creating accessor to an empty scratchpad buffer. We've addressed this in the latest commit